File: lab spring25/python a5/taskE.py
import sys
sys.stdin = open("input.txt","r")
# sys.setrecursionlimit(2*100000+5)

cycle = False

def dfs(graph,vst,rec,at):
    global cycle
    vst[at] = 1
    rec[at] = True
    for u in graph[at]:
        if vst[u] == 0:
            dfs(graph,vst,rec,u)
        elif rec[u]: cycle=True
    rec[at] = False

def solve():
    global cycle
    cycle = False
    n,m = list(map(int,input().split()))
    graph = [set() for _ in range(n+1)]
    for _ in range(m):
        u,v = list(map(int,input().split()))
        graph[u].add(v)
    visited = [0]*(n+1)
    for u in range(1,n+1):
        if visited[u]==0: dfs(graph,visited,[False]*(n+1),u)
    print("YES"if(cycle) else "NO")

t = 1
t = int(input())
while t!=0:
    t-=1
    solve()

File: lab spring25/assignment5/taskE.cpp
#include <stdio.h>
#include <vector>
#include <queue>
#include <stack>
using namespace std;

#define rept(i,a,b) for(int i=a;i<b;i++)
#define vi vector<int>
#define vb vector<bool>
#define pqi priority_queue<int>
#define br printf("\n")
#define yes printf("YES\n")
#define no printf("NO\n")

void makeGraph(pqi *graph, int m){
    int u,v;
    while(m--){
        scanf("%d %d",&u,&v);
        graph[u].push(v);
    }
}

bool cyclic = false;

void dfs(pqi *graph,vb &vst, int n, int s){
    vst[s] = true;
    stack<int> stk; stk.push(s);
    vb rec(n+1,false); rec[s] = true;

    while (!stk.empty()){
        int at{stk.top()};
        bool picked = false;
        pqi &g{graph[at]};

        while (!g.empty()){
            int u{g.top()}; g.pop();
            if(rec[u]) {
                cyclic = true;
                return;
            }
            if(!vst[u]){
                vst[u] = true;
                stk.push(u);
                picked = true;
                rec[u] = true;
                break;
            }
        }

        if(!picked){
            rec[stk.top()]=false;
            stk.pop();
        }
    }
}

void solve(){
    int n,m;
    scanf("%d %d",&n,&m);
    pqi graph[n+1];
    makeGraph(graph,m);
    cyclic = false;
    vb vst(n+1,false);
    for(int i=1; i<=n && !cyclic; i++) if(!vst[i]) dfs(graph,vst,n,i);
    cyclic?yes:no;
}

int main(){
    freopen("input.txt","r",stdin);
    int t{1};
    scanf("%d",&t);
    while(t--) solve();
}